keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,374 @@
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet and a new hint provider of Topology Manager for memory and hugepage. It provides NUMA node affinity to the Topology Manager to guarantee NUMA aware memory consumption for a container in the Guaranteed QoS Pod.
+
+When Guaranteed QoS Pod admission is requested to kubelet, Topology Manager asks NUMA node affinity of memory and hugepage to Memory Manager. Then Memory Manager calculates NUMA node affinity based on internal database(`Memory Map`) and sends the affinity back to the Topology Manager to help the Topology Manager to figure out appropriate NUMA node where adequate resources are available for a given container.
+
+Note:
+- Memory Manager only support single NUMA node affinity, because Linux kernel doesn`t have such mechanism to reserve memory or set memory limit per NUMA node.
+
+Then, while the container is being deployed, Memory Manager reserves requested amount of memory on the internal database(`Memory Map`). Then it sets [CPUSET enforcement](#cpuset-enforcement) to consume memory from designated NUMA node, that Topology Manager selected.
+
+Memory Manager guarantees that the managed container can consume reserved memory on the designated NUMA node. For this, Memory Manager depends on [OOM score enforcement](#oom-score-enforcement) of Kubelet.
+
+Consequently, user can expect Topology Manager(with single-numa-node policy) coordinates compute resources along with memory, hugepage to the same NUMA node.
+
+More details of this component are listed below.
+
+#### The Concept of Memory Map
+
+The Memory Manager has internal database(`Memory Map`), it is used to calcuate NUMA node affinity and store the information of reserved memory. Memory Manager perceives conventional memory and each hugepage size as different memory type to calculate topology hint precisely.
+
+- For an example, on the node which has hugepages-1Gbi and hugepages-2Mi, the number of memory type will be three : conventional memory, hugepages-1Gi, hugepages-2Mi.
+
+The Memory Map is the collection of Memory Table for each NUMA node and memory type. Memory Manager generates this Memory Map on its initialize sequence.
+Below figures show how memory map is consisted, in golang style.
+
+```go
+Type MemoryTable struct {
+    TotalMemSize uint64
+    PreReserved  uint64
+    Allocatable  uint64
+    Reserved     uint64
+    Free         uint64
+}
+
+type MemoryMap map[NUMBER_OF_NUMA_NODES]map[MEMORY_TYPES]MemoryTable
+```
+_Figure: The construction of memory map._
+
+- For a brief example, on a two-socket machine with one size of pre-allocated hugepages, Memory Manager will generates four memory tables(two of NUMA nodes * two of memory types).
```

Comment:
Do you implicitly or explicitly assume that the underlying hardware always runs with sub-NUMA clustering disabled (IOW that it looks like there was only one NUMA node per socket) ?

Reply:
I thought there will be no issue from SBC
As I know only platinum line support SNC isnt it?
and, once it enabled on bios, kernel exposes two(right?) NUMA node per socket with different NUMA index.

Memory Manager will be dealing with NUMA node index that parsed from cAdvisor.
If SNC is enabled, MM will perceive two NUMA nodes from one socket.

Reply:
> I thought there will be no issue from SBC
> As I know only platinum line support SNC isnt it?
> and, once it enabled on bios, kernel exposes two(right?) NUMA node per socket with different NUMA index.
> 
> Memory Manager will be dealing with NUMA node index that parsed from cAdvisor.
> If SNC is enabled, MM will perceive two NUMA nodes from one socket.

I didn't try to imply there would be problems, I just wanted to know whether you have a 1 node / socket assumption or not.

Reply:
Linux will present it as two NUMA nodes, and as far as I know Kubernetes only cares about NUMA nodes and not physical sockets.